visiont/scripts/train.py

```diff
@@ -0,0 +1,265 @@
+        param.requires_grad = False
+
+    def update_target():
+        update(target, online, 0.99)
+
+    # In addition to projection heads,
+    # The online network has predictor.
+    # predictor = mlp(256, 4096, 256)
+    predictor = mlp(768, 4096, 768)
+
+    # Move everything to devices
+
+    online = online.to(device)
+    online = nn.DataParallel(online)
+
+    predictor = predictor.to(device)
+    predictor = nn.DataParallel(predictor)
+
+    target = target.to(device)
+    target = nn.DataParallel(target)
+
+    def criterion(x, y):
+        x = nn.functional.normalize(x, dim=-1)
+        y = nn.functional.normalize(y, dim=-1)
+        return 2 - 2 * (x * y).sum(dim=-1)
+
+    lr = 1e-5
+
+    # Online and predictor learns, target gets assigned moving average of online network's weights.
+    # optimizer = torch.optim.Adam(list(online.parameters()) + list(predictor.parameters()), lr=lr)
+    optimizer = torch.optim.SGD(
+        list(online.parameters()) + list(predictor.parameters()), lr=lr
+    )
+
+    # Warmup Adam, he cold
+    def adjust_learning_rate(optimizer, step, lr):
+        for g in optimizer.param_groups:
+            g["lr"] = min(1, (step + 1) / 1000) * lr
+
+    step = 0
+
+    for epoch in range(100):
+        running = 0
+
+        progress = tqdm(loader, desc=f"Epoch {epoch+1}", unit="batch")
+
+        for idx, (inputs1, inputs2) in enumerate(progress):
+            assert inputs1.size() == inputs2.size()
+
+            # Overlap data transfers to gpus, pinned memory
+            inputs1 = inputs1.to(device, non_blocking=True)
+            inputs2 = inputs2.to(device, non_blocking=True)
+
+            adjust_learning_rate(optimizer, step, lr)
+
+            optimizer.zero_grad()
+
+            # Target network is in eval mode and does not
+            # require grads, forward no grad ctx to be sure
+            with torch.no_grad():
+                labels1 = target(inputs1).detach()
+                labels2 = target(inputs2).detach()
+
+            outputs1 = predictor(online(inputs1))
+            outputs2 = predictor(online(inputs2))
+
+            # Symmetrize the loss, both transformations
+            # go through both networks, one at a time
+            loss = criterion(outputs1, labels2) + criterion(outputs2, labels1)
+            loss = loss.mean()
+
+            loss.backward()
+
+            # Transformers need their nails clipped
+            # Harsimrat := best comment of 2020
+            nn.utils.clip_grad_norm_(online.parameters(), 1)
+
+            optimizer.step()
+
+            # After training the online network, we transfer
+            # a weighted average of the weights to the target
+            # Only apply warmup criterion on epoc == 0
+            # Update target network so if "lags" the online network
+            if idx > warmup_steps and epoch == 0:
+                update_target()
+            elif idx % target_update_steps == 0 and epoch > 0:
+                update_target()
```

Comment:
Why do we work with steps, but then special case the first epoch? Can warmup not work on steps alone?

The step var counts the total ("global") steps anyway.

Reply:
idx is local to the epoch. Warmup is only applied to first epoch